[PyTorch] [Model Tracer] Use c10::Synchronized<T> abstraction for mutex protected data

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73408
* #73407

Use the newly introduced `c10::Synchronized<T>` abstraction to enforce the use use of a mutex/lock on variables that should be protected for concurrent reads/modifications.

If reviewers think that this change is okay, I'll go ahead and change the other users in model tracer. This change is meant to show reviewers what the new code will look like.

Differential Revision: [D34465419](https://our.internmc.facebook.com/intern/diff/D34465419/)

dhruvbird committed:
[PyTorch] [Model Tracer] Use c10::Synchronized<T> abstraction for mutex protected data

Use the newly introduced `c10::Synchronized<T>` abstraction to enforce the use use of a mutex/lock on variables that should be protected for concurrent reads/modifications.

If reviewers think that this change is okay, I'll go ahead and change the other users in model tracer. This change is meant to show reviewers what the new code will look like.

Differential Revision: [D34465419](https://our.internmc.facebook.com/intern/diff/D34465419/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] [Model Tracer] Use c10::Synchronized<T> abstraction for mutex protected data"


Use the newly introduced `c10::Synchronized<T>` abstraction to enforce the use use of a mutex/lock on variables that should be protected for concurrent reads/modifications.

If reviewers think that this change is okay, I'll go ahead and change the other users in model tracer. This change is meant to show reviewers what the new code will look like.

Differential Revision: [D34465419](https://our.internmc.facebook.com/intern/diff/D34465419/)

[ghstack-poisoned]

cccclai commented:
I'm curious about the motivation of this implementation. Do we see regression somewhere? Or is there any use case requiring similar utility function?

dhruvbird commented:
> I'm curious about the motivation of this implementation. Do we see regression somewhere? Or is there any use case requiring similar utility function?

There was a bug that @ezyang reported that was fixed by adding mutexes around global data structures during tracing. However, that was a band-aid since it fixed the problem, but use usage is a bit ugly and requires users to be "Aware" of what needs to be protected by what mutex. This change adds an abstraction that takes the guess-work/risk out that for future developers.

dhruvbird committed:
Update on "[PyTorch] [Model Tracer] Use c10::Synchronized<T> abstraction for mutex protected data"


Use the newly introduced `c10::Synchronized<T>` abstraction to enforce the use use of a mutex/lock on variables that should be protected for concurrent reads/modifications.

If reviewers think that this change is okay, I'll go ahead and change the other users in model tracer. This change is meant to show reviewers what the new code will look like.

Differential Revision: [D34465419](https://our.internmc.facebook.com/intern/diff/D34465419/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch] [Model Tracer] Use c10::Synchronized<T> abstraction for mutex protected data"


Use the newly introduced `c10::Synchronized<T>` abstraction to enforce the use use of a mutex/lock on variables that should be protected for concurrent reads/modifications.

If reviewers think that this change is okay, I'll go ahead and change the other users in model tracer. This change is meant to show reviewers what the new code will look like.

Differential Revision: [D34465419](https://our.internmc.facebook.com/intern/diff/D34465419/)

[ghstack-poisoned]
